File: solutions/csharp/forth/2/StackHelpers.cs
public static class StackHelpers
{
    public static Context DuplicateLast(Context context)
    {
        _ = context ?? throw new ArgumentNullException(nameof(context));
        return context.WithStack(
            stack =>
            {
                int last = stack.RequireAtLeastOneValue()
                    .LastOrDefault();
                stack.Add(last);
            }
        );
    }

    public static Context DropLast(Context context)
    {
        _ = context ?? throw new ArgumentNullException(nameof(context));
        return context.WithStack(
            stack =>
                _ = stack.RequireAtLeastOneValue()
                    .Pop()
        );
    }

    public static Context SwapLast(Context context)
    {
        _ = context ?? throw new ArgumentNullException(nameof(context));
        return context.WithStack(
            stack =>
            {
                int lastValue = stack.RequireAtLeastTwoValues()
                    .Pop();
                int penultimateValue = stack.Pop();
                stack.Add(lastValue);
                stack.Add(penultimateValue);
            }
        );
    }

    public static Context PenultimateValueCopy(Context context)
    {
        _ = context ?? throw new ArgumentNullException(nameof(context));
        return context.WithStack(
            stack =>
            {
                int penultimateValue = stack.RequireAtLeastTwoValues()[^2];
                stack.Add(penultimateValue);
            }
        );
    }
}

File: solutions/java/forth/1/src/main/java/StackHelpers.java
import java.util.List;

public final class StackHelpers {
    public static Context duplicateLast(Context context) {
        requireContext(context);
        return context.withStack(stack -> {
            int last = ListUtilities.requireAtLeastOneValue("Duplicating", stack)
                                    .get(stack.size() - 1);
            stack.add(last);
        });
    }

    public static Context dropLast(Context context) {
        requireContext(context);
        return context.withStack(stack -> {
            final List<Integer> values = ListUtilities.requireAtLeastOneValue("Dropping", stack);
            ListUtilities.pop(values);
        });
    }

    public static Context swapLast(Context context) {
        requireContext(context);
        return context.withStack(stack -> {
            final List<Integer> values = ListUtilities.requireAtLeastTwoValues("Swapping", stack);
            int lastValue = ListUtilities.pop(values);
            int penultimateValue = ListUtilities.pop(stack);
            stack.add(lastValue);
            stack.add(penultimateValue);
        });
    }

    public static Context penultimateValueCopy(Context context) {
        requireContext(context);
        return context.withStack(stack -> {
            List<Integer> validatedStack = ListUtilities.requireAtLeastTwoValues("Overing", stack);
            int penultimateValue = validatedStack.get(validatedStack.size() - 2);
            stack.add(penultimateValue);
        });
    }

    private static void requireContext(Context context) {
        if (context == null) {
            throw new NullPointerException("context");
        }
    }
}